File: gtci/12-tree-depth-first-search/python3/7-tree-diameter.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findDiameter(self, root):
        if root is None:
            return 0

        self.diameter = 0
        self.longestPath(root)

        return self.diameter

    def longestPath(self, root):
        if root is None:
            return 0

        longest_left_path = self.longestPath(root.left)
        longest_right_path = self.longestPath(root.right)

        self.diameter = max(self.diameter, longest_left_path + longest_right_path + 1)

        return max(longest_left_path, longest_right_path) + 1


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    assert sol.findDiameter(root) == 5

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    root.right.left.left = TreeNode(7)
    root.right.right.right = TreeNode(8)
    root.right.left.left.left = TreeNode(9)
    root.right.right.right.left = TreeNode(10)
    root.right.right.right.left.left = TreeNode(11)
    assert sol.findDiameter(root) == 8

    print("All test cases passed.")

File: gtci/12-tree-depth-first-search/cpp/7-tree-diameter.cpp
#include <algorithm>
#include <cassert>
#include <iostream>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  int diameter = 0;
  int findDiameter(TreeNode* root) {
    if (root == nullptr)
      return 0;
    
    longestPath(root);

    return this->diameter;
  }

  int longestPath(TreeNode* node) {
    if (node == nullptr)
      return 0;
    
    int longestLeftPath = longestPath(node->left);
    int longestRightPath = longestPath(node->right);

    this->diameter = std::max(this->diameter, longestLeftPath + longestRightPath + 1);

    return std::max(longestLeftPath, longestRightPath) + 1;
  }
};

int main() {
    Solution sol;
    
    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    root1->left->left = new TreeNode(4);
    root1->right->left = new TreeNode(5);
    root1->right->right = new TreeNode(6);
    assert(sol.findDiameter(root1) == 5);
    
    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    root2->right = new TreeNode(3);
    root2->right->left = new TreeNode(5);
    root2->right->right = new TreeNode(6);
    root2->right->left->left = new TreeNode(7);
    root2->right->right->right = new TreeNode(8);
    root2->right->left->left->left = new TreeNode(9);
    root2->right->right->right->left = new TreeNode(10);
    root2->right->right->right->left->left = new TreeNode(11);
    assert(sol.findDiameter(root2) == 8);
    
    cout << "All test cases passed." << endl;

    delete root1;
    delete root2;
    
    return 0;
}
